src/index.js

```diff
@@ -1,102 +1,144 @@
+   * @returns {Connection} An upgraded Connection
+   */
+  async dial (ma, options = {}) {
+    const stream = await this._connect(ma, options)
+    const maConn = toConnection({
+      stream,
+      remoteAddr: ma
+    }, options)
+    log('new outbound connection %s', maConn.remoteAddr)
+    const conn = await this._upgrader.upgradeOutbound(maConn)
+    log('outbound connection %s upgraded', maConn.remoteAddr)
+    return conn
   }
 
-  dial (ma, options, callback) {
-    if (isFunction(options)) {
-      callback = options
-      options = {}
+  /**
+   * @private
+   * @param {Multiaddr} ma
+   * @param {object} options
+   * @param {AbortSignal} options.signal Used to abort dial requests
+   * @returns {Promise<DuplexIterableStream>} Resolves a duplex iterable stream with the target peer
+   */
+  _connect (ma, options = {}) {
+    if (options.signal && options.signal.aborted) {
+      throw new AbortError()
     }
 
-    callback = once(callback || noop)
-
-    const server = this.connections[getServerForAddress(ma)]
-    const conn = new Connection()
-
-    server._dial(ma).then(_conn => {
-      conn.setInnerConn(_conn)
-      callback()
-    }, callback)
+    const server = this.listeners[getServerForAddress(ma)]
 
-    conn.getObservedAddrs = (callback) => {
-      return callback(null, [ma])
+    if (!server) {
+      throw errCode(new Error(messages.NO_LISTENING_SERVER), codes.NO_LISTENING_SERVER)
     }
 
-    return conn
+    return server._dial(ma, options)
   }
 
+  /**
+   * Creates a stardust listener. The provided `handler` function will be called
+   * anytime a new incoming Connection has been successfully upgraded via
+   * `upgrader.upgradeInbound`.
+   * @param {object} [options]
+   * @param {number} [options.discoveryInterval]
+   * @param {function (Connection)} handler
+   * @returns {Listener} A stardust listener
+   */
   createListener (options, handler) {
-    if (isFunction(options)) {
+    if (typeof options === 'function') {
       handler = options
       options = {}
     }
 
-    handler = handler || noop
-
-    return new Listener(this, handler)
+    return new Listener({
+      handler,
+      upgrader: this._upgrader,
+      client: this,
+      options
+    })
   }
 
+  /**
+   * Takes a list of `Multiaddr`s and returns only valid Stardust addresses
+   * @param {Multiaddr[]} multiaddrs
+   * @returns {Multiaddr[]} Valid Stardust multiaddrs
+   */
   filter (multiaddrs) {
-    if (!Array.isArray(multiaddrs)) {
-      multiaddrs = [multiaddrs]
-    }
+    multiaddrs = Array.isArray(multiaddrs) ? multiaddrs : [multiaddrs]
 
     return multiaddrs.filter((ma) => {
-      if (includes(ma.protoNames(), 'p2p-circuit')) {
+      if (ma.protoCodes().includes(CODE_CIRCUIT)) {
         return false
       }
 
-      if (includes(ma.protoNames(), 'p2p-stardust')) {
+      if (ma.protoNames().includes('p2p-stardust')) {
         return true
       }
 
-      // return mafmt.Stardust.matches(ma)
```

Comment:
I suppose it would be better if we fix & update mafmt to match the addresses correctly. I only commented this out and added the other snippet as a quickfix some time ago.

Reply:
Yes, I initially was thinking the same. 

We should postpone this to after deciding wether stardust should be usable on reliable addresses, or just ws/wss, or even just wss. I forgot to open an issue to discuss this, but I will open it now and we can make this happen in a followup PR after a decision is made.

Reply:
Ok, sounds good. My opinion is that (since the security concern is mainly DOS) it's not wrong to allow it, yet there should be a way we make it clear to the user that it might be an issue that way.

Or stardust could check if the address has an encryption layer (for ex wss) and add one if necessary (using any of the libp2p crypto layers). We could accomplish this by adding secio+plaintext to the server and then using either a plaintext or secio upgrader on the client.

Reply:
can you post that on #22 ? It will be easier to discuss from there